File: app_java_wiki_chat/spring-backend/src/main/java/com/platform/securelearning/chat/ChatHistoryService.java
package com.platform.securelearning.chat;

import com.platform.securelearning.chat.dto.ChatHistoryEntry;
import com.platform.securelearning.crypto.EncryptedPayload;
import com.platform.securelearning.crypto.EncryptionService;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.List;
import org.springframework.stereotype.Service;

/** Encrypts each turn before it touches the database, decrypts on the way back out (SR-4). */
@Service
public class ChatHistoryService {

    private final ChatMessageRepository repository;
    private final EncryptionService encryptionService;

    public ChatHistoryService(ChatMessageRepository repository, EncryptionService encryptionService) {
        this.repository = repository;
        this.encryptionService = encryptionService;
    }

    public void save(String roomCode, String senderSessionId, String plaintext) {
        EncryptedPayload payload = encryptionService.encrypt(plaintext.getBytes(StandardCharsets.UTF_8));
        repository.save(new ChatMessage(roomCode, senderSessionId, payload.ciphertext(), payload.nonce(), Instant.now()));
    }

    public List<ChatHistoryEntry> history(String roomCode) {
        return repository.findByRoomCodeOrderByCreatedAtAsc(roomCode).stream()
                .map(message -> new ChatHistoryEntry(
                        message.getSenderSessionId(),
                        new String(encryptionService.decrypt(
                                new EncryptedPayload(message.getCiphertext(), message.getNonce())),
                                StandardCharsets.UTF_8),
                        message.getCreatedAt()))
                .toList();
    }
}

File: app02_csharp_wiki_chat/aspnet-backend/SecureLearning.Backend/Chat/ChatHistoryService.cs
using System.Text;
using SecureLearning.Backend.Crypto;

namespace SecureLearning.Backend.Chat;

public sealed record ChatHistoryEntry(string SenderSessionId, string Text, DateTimeOffset CreatedAt);

/// <summary>Encrypts each turn before it reaches storage, decrypts on the way back out (SR-4).</summary>
public sealed class ChatHistoryService(IChatMessageRepository repository, IEncryptionService encryption)
{
    public void Save(string roomCode, string senderSessionId, string plaintext)
    {
        var payload = encryption.Encrypt(Encoding.UTF8.GetBytes(plaintext));
        repository.Save(new ChatMessage(roomCode, senderSessionId, payload.Ciphertext,
            payload.Nonce, DateTimeOffset.UtcNow));
    }

    public IReadOnlyList<ChatHistoryEntry> History(string roomCode) =>
        repository.FindByRoomOrdered(roomCode)
            .Select(m => new ChatHistoryEntry(
                m.SenderSessionId,
                Encoding.UTF8.GetString(encryption.Decrypt(new EncryptedPayload(m.Ciphertext, m.Nonce))),
                m.CreatedAt))
            .ToList();
}
